### PR TITLE
AP_RCProtocol: Use a data structure to hold serial configurations for auto-detection

### PR DESCRIPTION
**Pull request by peterbarker**
More data structure, less code.

I think this makes it clearer as to what is going on, and makes it easier to add new serial configurations.

### CODE REVIEW

File: libraries/AP_RCProtocol/AP_RCProtocol.cpp [hunk 1]
```diff
@@ -238,6 +238,27 @@ void AP_RCProtocol::process_handshake( uint32_t baudrate)
...
+    uart->configure_parity(parity);
+    uart->set_stop_bits(stop_bits);
+    uart->set_options(uart->get_options() & ~masked_out_options);
+    uart->set_options(uart->get_options() | masked_in_options);
+    uart->begin(baud, 128, 128);
+}
+
+static const AP_RCProtocol::SerialConfig serial_configs[] {
+    // inverted and uninverted 115200 8N1:
+    { 115200, 0, 1, 0, AP_HAL::UARTDriver::OPTION_RXINV },
```
Comment (andyp1per) [inline libraries/AP_RCProtocol/AP_RCProtocol.cpp:252]: Can you turn parity and stop bits into constants

Comment (peterbarker) [reply libraries/AP_RCProtocol/AP_RCProtocol.cpp]:
```
    { 115200, AP_RCProtocol::ParityBits::ZERO, AP_RCProtocol::StopBits::ONE, 0, AP_HAL::UARTDriver::OPTION_RXINV },
    { 115200, AP_RCProtocol::ParityBits::ZERO, AP_RCProtocol::StopBits::ONE, AP_HAL::UARTDriver::OPTION_RXINV, 0 },
    // SBUS settings, even parity, 2 stop bits:
    { 100000, AP_RCProtocol::ParityBits::TWO, AP_RCProtocol::StopBits::TWO, AP_HAL::UARTDriver::OPTION_RXINV, 0 },
    // CrossFire:
    { 416666, AP_RCProtocol::ParityBits::ZERO, AP_RCProtocol::StopBits::ONE, 0, AP_HAL::UARTDriver::OPTION_RXINV },
```

yeah, that didn't work out :-)

Do you think a comment will cover it?

```
static const AP_RCProtocol::SerialConfig serial_configs[] {
    // BAUD PRTY STOP MASKED-IN MASKED-OUT
    // inverted and uninverted 115200 8N1:
    { 115200,  0,   1, 0, AP_HAL::UARTDriver::OPTION_RXINV },
    { 115200,  0,   1, AP_HAL::UARTDriver::OPTION_RXINV, 0 },
    // SBUS settings, even parity, 2 stop bits:
    { 100000,  2,   2, AP_HAL::UARTDriver::OPTION_RXINV, 0 },
    // CrossFire:
    { 416666,  0,   1, 0, AP_HAL::UARTDriver::OPTION_RXINV },
};
```

Comment (andyp1per) [reply libraries/AP_RCProtocol/AP_RCProtocol.cpp]: Yep fine :)

**Comment by peterbarker** [comment]
I've tested this with
 - RCIN on a MiniPix
 - timer-capture on MatekH743
 - uart on MatekH743
 - RCIN on a CubeBlack
 - Spektrum input on a CubeBlack

**Comment by peterbarker** [comment]
```
Binary           text            data         bss          total
---------------  --------------  -----------  -----------  --------------
arduplane        -40 (-0.0023%)  0 (0.0000%)  0 (0.0000%)  -40 (-0.0020%)
antennatracker   -40 (-0.0030%)  0 (0.0000%)  0 (0.0000%)  -40 (-0.0025%)
arducopter-heli  -40 (-0.0023%)  0 (0.0000%)  0 (0.0000%)  -40 (-0.0020%)
arducopter       -40 (-0.0023%)  0 (0.0000%)  0 (0.0000%)  -40 (-0.0020%)
ardurover        -40 (-0.0026%)  0 (0.0000%)  0 (0.0000%)  -40 (-0.0022%)
blimp            -40 (-0.0032%)  0 (0.0000%)  0 (0.0000%)  -40 (-0.0026%)
ardusub          -40 (-0.0026%)  0 (0.0000%)  0 (0.0000%)  -40 (-0.0022%)
```

File: libraries/AP_RCProtocol/AP_RCProtocol.cpp [hunk 2]
```diff
@@ -238,6 +238,28 @@ void AP_RCProtocol::process_handshake( uint32_t baudrate)
   check for bytes from an additional uart. This is used to support RC
   protocols from SERIALn_PROTOCOL
  */
+void AP_RCProtocol::SerialConfig::apply_to_uart(AP_HAL::UARTDriver *uart) const
+{
+    uart->configure_parity(parity);
+    uart->set_stop_bits(stop_bits);
+    uart->set_options(uart->get_options() & ~masked_out_options);
```
Comment (tridge) [inline libraries/AP_RCProtocol/AP_RCProtocol.cpp:245]: calling set_options twice reconfigures the hw twice
a bool for RXINV would avoid this

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_RCProtocol/AP_RCProtocol.cpp (hunk 2 above)
Comment (peterbarker) [reply libraries/AP_RCProtocol/AP_RCProtocol.cpp]: Done

**Pull request merged**